sinabs/backend/dynapcnn/sinabs_edges_handler.py

```diff
@@ -0,0 +1,472 @@
+        key is unique dynapcnn layer ID, value is dict with nodes of the layer
+        Will be updated in-place.
+    edge: Tuple of 2 integers, indicating edge between two nodes in graph.
+        Edge source has to be within an existing entry of `dynapcnn_layer_info`.
+    node_2_layer_map (dict): Maps each node ID to the ID of the layer it is assigned to.
+        Will be updated in-place.
+    nodes_io_shapes (dict): Map from node ID to dict containing node's in- and output shapes
+    """
+    # Make sure both source (neuron layer) and target (weight layer) have been previously processed
+    try:
+        source_layer_idx = node_2_layer_map[edge[0]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[0])
+    try:
+        destination_layer_idx = node_2_layer_map[edge[1]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[1])
+
+    # Add new destination
+    output_shape = nodes_io_shapes[edge[0]]["output"]
+    layer_info = dynapcnn_layer_info[source_layer_idx]
+    layer_info["destinations"].append(
+        {
+            "pooling_ids": [],
+            "pooling_modules": [],
+            "destination_layer": destination_layer_idx,
+            "output_shape": output_shape,
+        }
+    )
+
+    # Add output shape of this layer to input shapes of destination
+    dynapcnn_layer_info[destination_layer_idx]["inferred_input_shapes"].add(
+        output_shape
+    )
+
+
+def set_pooling_layer_destination(
+    dynapcnn_layer_info: Dict[int, Dict],
+    edge: Edge,
+    node_2_layer_map: Dict[int, int],
+    nodes_io_shapes: Dict[int, Dict[str, Tuple[Size, Size]]],
+) -> None:
+    """Set destination layer with pooling.
+
+    Parameters
+    ----------
+    dynapcnn_layer_info: Dict with one entry for each future dynapcnn layer.
+        key is unique dynapcnn layer ID, value is dict with nodes of the layer
+        Will be updated in-place.
+    edge: Tuple of 2 integers, indicating edge between two nodes in graph.
+        Edge source has to be within an existing entry of `dynapcnn_layer_info`.
+    node_2_layer_map (dict): Maps each node ID to the ID of the layer it is assigned to.
+        Will be updated in-place.
+    nodes_io_shapes (dict): Map from node ID to dict containing node's in- and output shapes
+    """
+    # Make sure both source (pooling layer) and target (weight layer) have been previously processed
+    try:
+        source_layer_idx = node_2_layer_map[edge[0]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[0])
+    try:
+        destination_layer_idx = node_2_layer_map[edge[1]]
+    except KeyError:
+        raise UnmatchedNode(edge, edge[1])
+
+    # Find current source node within destinations
+    layer_info = dynapcnn_layer_info[source_layer_idx]
+    matched = False
+    for destination in layer_info["destinations"]:
+        if destination["pooling_ids"][-1] == edge[0]:
+            if destination["destination_layer"] is not None:
```

Comment:
Maybe I understood this wrong but shouldn't we check `if destination["destination_layer"] == destination_layer_idx`?

Reply:
Maybe I understand now: this will be the case where the destinations dict has been set by a "neuron-weight" edge, meaning that the neuron layer itself projects to another dynapcnnlayer X without pooling, but it also goes through some pooling before targeting another dynapcnnlayer Y. 

So the `layer_info["destinations"].append(destination)` line  copies the destination dict initialized by the `set_neuron_layer_destination` call but updates the `destination_layer` and `output_shape` to set the pooling.

Is this the case @bauerfe ?

Reply:
Almost. The destination dict has not been initialized by `set_neuron_layer_destination` but by `set_pooling_layer_destination`, and the final pooling node is the same as `edge[0]`. Otherwise the `if` statement where it compares the last "pooling_id" with `edge[0]` would be False.

Reply:
I still don't see it cuz in the sequence of function calls that `collect_dynapcnn_layer_info` makes the only one that sets the `destination_layer` in the dict of destinations is the function `set_neuron_layer_destination`, so it will only not be `None` when the neuron node (from when a **neuron-weight** edge was processed) associated with the pooling node being processed by `set_pooling_layer_destination` (processing **pooling-weight** nodes), so the pooling node in this edge must be in the same DynapcnnLayer that the previously mentioned neuron node is.

Can we go over this together at the call later today?

Reply:
I think my previous comment caused more confusion because I confused `add_pooling_to_entry` with `set_pooling_layer_destination`.

Of course we can discuss this in our call. I will still try to add some clarification, so that it can be looked up at a later point and possibly be used to improve the readability of the code:

Both functions `set_neuron_layer_destination` and `add_pooling_to_entry` add destination-dicts to a layer info. The dicts added by `set_neuron_layer_destination` will never have destination layer `None`, and they will always have empty lists as `pooling_ids`. Such entries will fail both if-conditions in ll. 402f. 

On the other hand `add_pooling_to_entry` will add a dict with non-empty `pooling_ids`, but `None` as `destination_layer`, because it is not yet defined at this point. Later on, when the edge is discovered that connects the pooling node of that destination-dict to another layer, `set_pooling_layer_destination` will be called to enter that `destination_layer` . However, it can happen that the same pooling node is connected to multiple other dynapcnn layers. In that case a new destination dict has to be created for each, to prevent the function from overwriting the same value each time.

Reply:
Actually I think I see it, your comment "_this is done if there are multiple edges to another dynapcnn layer destination from a single pooling node_" made sense now looking at the code. Let's go over it in the call quickly anyway to make sure my understanding is correct.

Reply:
You are right, it's redundant.
It is just needed in the case where the destination does not have to be copied. 
The question is, does it make sense to add complexity to the code by considering the two cases separately?